x-pack/libbeat/management/unit.go

```diff
@@ -0,0 +1,410 @@
+			continue
+		}
+
+		if cfgName := expectedCfg.GetName(); cfgName != "" {
+			id := fmt.Sprintf("%s.[%d]", cfgName, idx)
+			streamIDs[idx] = id
+			streamStates[id] = streamState
+			continue
+		}
+
+		id := fmt.Sprintf("%s.[%d]", expectedCfg.GetId(), idx)
+		streamIDs[idx] = id
+		streamStates[id] = streamState
+	}
+
+	return streamStates, streamIDs
+}
+
+// newAgentUnit creates a new agentUnit. In case the supplied client.Unit is of type
+// client.UnitTypeInput it initializes the streamStates with a unitState.Unknown
+func newAgentUnit(cu *client.Unit, log *logp.Logger) *agentUnit {
+	var (
+		streamStates map[string]unitState
+		streamIDs    []string
+	)
+
+	if cu.Type() == client.UnitTypeInput {
+		streamStates, streamIDs = getSteamStates(cu.Expected())
+	}
+
+	return &agentUnit{
+		clientUnit:   cu,
+		logger:       log,
+		streamIDs:    streamIDs,
+		streamStates: streamStates,
+	}
+}
+
+// RegisterAction registers action handler for this unit.
+func (u *agentUnit) RegisterAction(action client.Action) {
+	u.mtx.Lock()
+	defer u.mtx.Unlock()
+
+	if u.clientUnit == nil {
+		return
+	}
+
+	u.clientUnit.RegisterAction(action)
+}
+
+// UnregisterAction unregisters action handler with the client.
+func (u *agentUnit) UnregisterAction(action client.Action) {
+	u.mtx.Lock()
+	defer u.mtx.Unlock()
+
+	if u.clientUnit == nil {
+		return
+	}
+
+	u.clientUnit.UnregisterAction(action)
+}
+
+func (u *agentUnit) Expected() client.Expected {
+	u.mtx.Lock()
+	defer u.mtx.Unlock()
+
+	if u.clientUnit == nil {
+		return client.Expected{}
+	}
+
+	return u.clientUnit.Expected()
+}
+
+func (u *agentUnit) ID() string {
+	u.mtx.Lock()
+	defer u.mtx.Unlock()
+
+	if u.clientUnit == nil {
+		return ""
+	}
+
+	return u.clientUnit.ID()
+}
+
+// calcState calculates the current state of the unit.
+func (u *agentUnit) calcState() (status.Status, string) {
+	// for type output return the unit state directly as it has no streams
+	if u.clientUnit.Type() == client.UnitTypeOutput {
+		return u.input.state, u.input.msg
+	}
+
+	// if input state is not running return the input state
+	if u.input.state != status.Running {
+		return u.input.state, u.input.msg
+	}
+
+	// input state is marked as running, check the stream states
+	countFailed := 0
+	countDegraded := 0
+	for _, streamState := range u.streamStates {
+		switch streamState.state {
+		case status.Degraded:
+			countDegraded++
+		case status.Failed:
+			countFailed++
+		}
+	}
+
+	if countDegraded > 0 && countFailed > 0 {
+		return status.Failed, "Some streams are Failed and some are Degraded"
+	}
+
+	if countDegraded > 0 {
+		if countDegraded == len(u.streamStates) {
+			return status.Degraded, "All streams are Degraded"
+		} else {
+			return status.Degraded, "Some streams are Degraded"
+		}
+	}
+
+	if countFailed > 0 {
+		if countFailed == len(u.streamStates) {
+			return status.Failed, "All streams are Failed"
+		} else {
+			return status.Failed, "Some streams are Failed"
+		}
+	}
```

Comment:
Could we simplify this block by setting an initial `status.Healthy` and as we go through the streams we make the transition `Healthy->Degraded` when a degraded stream is encountered and `* -> Failed` when a failed stream is found? A message for both the degraded and failed status can be "%d streams degrade, %d streams failed"... Something like the snippet below:
```suggestion
	// input state is marked as running, check the stream states
	status := status.Running
	statusMessage := "Healthy"
	countFailed := 0
	countDegraded := 0
	for _, streamState := range u.streamStates {
		switch streamState.state {
		case status.Degraded:
			countDegraded++
			if status == status.Healthy {
				status= status.Degraded
			}
		case status.Failed:
			countFailed++
			status = status.Failed
		}
	}

	if countDegraded > 0 || countFailed > 0 {
		statusMessage = fmt.Sprintf("Out of %d streams, %d are failed, %d are degraded", len(u.streamStates), countFailed, countDegraded)
	}

	return status, statusMessage

```

Reply:
> Could we simplify this block by setting an initial `status.Healthy` and as we go through the streams we make the transition `Healthy->Degraded` when a degraded stream is encountered and `* -> Failed` when a failed stream is found? A message for both the degraded and failed status can be "%d streams degrade, %d streams failed"... Something like the snippet below:

Although the code snippet you suggested wasn't compiling I adjusted and it is now in. Feel free to unresolve if I missed anything during the adjustments